packages/libp2p/src/dcutr/dcutr.ts

```diff
@@ -214,9 +253,67 @@ export class DefaultDCUtRService implements Startable {
     }
   }
 
+  /**
+   * This function encapsulates the Simultaneous Connect step (#5) of the [DCUtR protocol](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol)
+   * @param relayedConnection
+   * @param multiaddrs
+   * @param signal
+   */
+  async handleSimultaneousConnect (relayedConnection: Connection, multiaddrs: Multiaddr[], rtt: number, signal: AbortSignal): Promise<void> {
+    logB('waiting for half RTT')
+    // ..and starts a timer for half the RTT measured from the time between
+    // sending the initial Connect and receiving the response
+    // TODO: Find a way to simulate network latency for tests, because rtt is always <= 4ms
+    await delay(rtt / 2)
+
+    const tcpMultiaddrs = multiaddrs.filter(ma => ma.protoNames().includes('tcp'))
+    const quicMultiaddrs = multiaddrs.filter(ma => ma.protoNames().includes('quic'))
+
+    logB('dialing', multiaddrs)
+    const connectAttempts = []
+    if (tcpMultiaddrs.length > 0) {
+      connectAttempts.push(this.simultaneousConnectTCP(tcpMultiaddrs, signal))
+    }
+    if (quicMultiaddrs.length > 0) {
+      connectAttempts.push(this.simultaneousConnectQuic(quicMultiaddrs, signal))
+    }
+    await Promise.any(connectAttempts)
```

Comment:
When QUIC and TCP addresses are present this won't cancel outstanding dial attempts.

Normally the connection manager would do this when passing multiple multiaddrs, but there are two separate calls to `.openConnection` here.

Instead we should pass a shared `AbortSignal` in and abort it after `Promise.any` returns, this will cancel any outstanding dials.

Reply:
https://github.com/libp2p/js-libp2p/pull/1932/commits/99f23d6e98385d8bad883ddf18c5db57fca52a6b